Add UnliftIO.Process/UnliftIO.Environment

TerrorJack opened the pull request:
I'm adding `UnliftIO.Process` which lifts `System.Process`. This adds a dependency on `process` but shouldn't pose a burden since it's a boot lib.

`UnliftIO.Environment` lifts `System.Environment`.

snoyberg commented on unliftio/src/UnliftIO.hs:
```diff
@@ -4,10 +4,12 @@ module UnliftIO
   , module UnliftIO.IO
   , module UnliftIO.IORef
   , module UnliftIO.MVar
+  , module UnliftIO.Process
```
I'm concerned with exporting both of these modules from `UnliftIO` itself, since they're likely to conflict with names that downstream packages may want to use. In particular, I can imagine people wanting to use `Text`-based values for environment handling, and `rio` is already providing a `RIO.Process` module that is based on `typed-process` instead of `process`.

TerrorJack commented:
@snoyberg Then I'll recover the export list of the UnliftIO module later, and let people decide when to import my newly-added modules later.

TerrorJack commented:
@snoyberg Hi Michael, I've removed re-exports for newly added modules as you requested. I've also added `UnliftIO.Foreign` which lifts `Foreign`, so much for this PR for now.

TerrorJack commented:
Seems UnliftIO.Foreign needs a bit more tinkering, will fix later.

TerrorJack commented:
@snoyberg Hi Michael, I've fixed UnliftIO.Foreign and you can review and merge now. There are 3 added modules in this PR: `UnliftIO.Environment`, `UnliftIO.Foreign`, `UnliftIO.Process` which are all straightforward wrappers for things in `base`/`process`, and the modules are not re-exported by `UnliftIO`.

snoyberg commented on unliftio/src/UnliftIO/Foreign.hs:
```diff
@@ -0,0 +1,1348 @@
+-- @since 0.2.4.0
+{-# INLINE newCAString #-}
+newCAString :: MonadIO m => String -> m CString
+newCAString = liftIO . newCAString
```
I think this is an infinite loop.

TerrorJack replied:
Fixed.

snoyberg commented:
Perfect, thanks!
